java/checkstyle/checkstyle.xml

```diff
@@ -0,0 +1,238 @@
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one or more
+  ~ contributor license agreements.  See the NOTICE file distributed with
+  ~ this work for additional information regarding copyright ownership.
+  ~ The ASF licenses this file to You under the Apache License, Version 2.0
+  ~ (the "License"); you may not use this file except in compliance with
+  ~ the License.  You may obtain a copy of the License at
+  ~
+  ~    http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<!DOCTYPE module PUBLIC
+          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
+          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
+
+<!--
+    Checkstyle configuration that checks the Google coding conventions from Google Java Style
+    that can be found at https://google.github.io/styleguide/javaguide.html.
+
+    Checkstyle is very configurable. Be sure to read the documentation at
+    http://checkstyle.sf.net (or in your downloaded distribution).
+
+    To completely disable a check, just comment it out or delete it from the file.
+
+    Authors: Max Vetrenko, Ruslan Diachenko, Roman Ivanov.
```

Comment:
So this is pretty annoying. checkstyle is `LGPL` which means this file cannot simply be copy-and-pasted from the checkstyle project (https://www.apache.org/legal/resolved.html#category-x) . Can we find a checkstyle.xml that implements the Google Java style from somewhere else that is ASF-compatible for inclusion here?

Reply:
Ahhh, anonying.

How about https://github.com/GoogleCloudPlatform/google-cloud-intellij/blob/master/google_checks.xml ?

Reply:
Looks like that file was also copied from the checkstyle project. I am looking at git histories for these files to try to determine the IP origin.

Kudu has this file, but it looks like it also came directly out of checkstyle https://github.com/apache/kudu/blob/master/java/kudu_style.xml. cc @jdcryans

looks like we can use the one from Apache Beam, we might want to review that it's the style that we expect (e.g. import order, etc.)

https://github.com/apache/beam/blob/a6f460fe3b760aafbc748ae18956f0f2c1fedfad/sdks/java/build-tools/src/main/resources/beam/checkstyle.xml

Reply:
It seems most "google_checks.xml" on the internet are from checkstyle project. Although some ASF licensed projects are using it and/or re license it under ASF license? 

To be on the safe side, we could use the checkstyle from beam and make changes if we have to. I can look into that.

Reply:
I believe we decided it was OK because the checkstyle xml doesn't end up linked into any resulting binaries, etc -- similar how it's OK to rely on GPL or LGPL build tools in a project. I do recall finding some precedent in other Apache projects or legal resolutions, but can't find it now looking back at my email.

Reply:
I see. It looks like we can use the Beam file and avoid the issue altogether, but it would be good to have the definitive judgment from ASF legal

Reply:
I am actually leaning towards using google_checks.xml if possible. The beam file seems to be somewhat different of the google checks, which seems to be more widely used and probably is easier to find an IDE codestyle config for. 

Some other ASF project, such as Spark, also seems to be using google checks:

https://github.com/apache/spark/blob/master/dev/checkstyle.xml

@wesm Is it possible to consult with ASF legal?

Reply:
With both Spark and Kudu using the checkstyle as is, I think it's OK to include. If the issue comes up later we can revisit. Thanks for looking into this

Reply:
If we want to be pedantic on this issue, we should ensure that the file does not end up in the source tarballs. As we are using `git archive` to create the final tarballs, it should be sufficent to add an entry to `.gitattrbutes`:

```
/java/checkstyle/checkstyle.xml export-ignore
```

Reply:
Thank @xhochy! I added .gitattribtues